common/src/main/java/com/wynntils/handlers/inventory/InventoryHandler.java

```diff
@@ -0,0 +1,449 @@
+            case PICKUP -> {
+                if (slotNum < 0) { // Slot -999 -> tossed an item outside the GUI
+                    if (!heldStack.isEmpty()) {
+                        if (event.getMouseButton() == 0 || heldStack.getCount() == 1) {
+                            forceSyncLater(menu);
+                        }
+                        expect(new PendingInteraction.ThrowFromHeld(menu, heldStack.copy()));
+                    }
+                } else {
+                    ItemStack slotStack = menu.getSlot(slotNum).getItem();
+                    if (heldStack.isEmpty()) {
+                        if (!slotStack.isEmpty()) {
+                            expect(new PendingInteraction.PickUp(menu, slotNum, slotStack.copy()));
+                        }
+                    } else if (slotStack.isEmpty()) {
+                        if (event.getMouseButton() == 0 || heldStack.getCount() == 1) {
+                            forceSyncLater(menu);
+                        }
+                        expect(new PendingInteraction.Place(menu, slotNum, heldStack.copy()));
+                    } else if (areItemsSimilar(heldStack, slotStack)) {
+                        expect(new PendingInteraction.Place(menu, slotNum, heldStack.copy()));
+                    } else {
+                        expect(new PendingInteraction.Swap(menu, slotNum, heldStack.copy(), slotStack.copy()));
+                    }
+                }
+            }
+            case QUICK_MOVE -> {
+                ItemStack slotStack = menu.getSlot(slotNum).getItem();
+                if (!slotStack.isEmpty()) {
+                    expect(new PendingInteraction.Transfer(menu, slotNum, slotStack.copy()));
+                }
+            }
+            case THROW -> {
+                ItemStack slotStack = menu.getSlot(slotNum).getItem();
+                if (heldStack.isEmpty() && !slotStack.isEmpty()) {
+                    forceSyncLater(menu);
+                    expect(new PendingInteraction.ThrowFromSlot(menu, slotNum, slotStack.copy()));
+                }
+            }
+            case QUICK_CRAFT -> {
+                if (!heldStack.isEmpty()) {
+                    running = new RunningInteraction(menu, (event.getMouseButton() & 0x4) != 0, heldStack.copy());
+                }
+            }
+            case PICKUP_ALL -> {
+                if (!heldStack.isEmpty()) {
+                    expect(new PendingInteraction.PickUpAll(menu, heldStack.copy()));
+                }
+            }
+                // SWAP and CLONE are not supported as inventory operations on Wynncraft
+        }
+    }
+
+    @SubscribeEvent
+    public void onSetContents(ContainerSetContentEvent.Pre event) {
+        long now = System.currentTimeMillis();
+        int containerId = event.getContainerId();
+        List<ItemStack> newInv = event.getItems();
+        pending.removeIf(entry -> now >= entry.expiryTime || !entry.interaction.onContentUpdate(containerId, newInv));
```

Comment:
We have a TimedSet (or something like that) that removes items after certain amount of time. 

But I am not sure what you are using this pending list for. I can't see any reads from it?

Reply:
What kind of overhead is it you are afraid of? Performance-wise? I'd like to see us utilize our existing functionality as much as possible, instead of creating new similar functionality. And if the existing functionality is missing something, it is better to improve it.

This is of course always a judgement call, and I understand why this particular instance is really a borderline case. Nevertheless, using `TimedSet` would make it show directly even at the declaration of `pending` that it is a, well, timed set. And it would make `PendingInteractionEntry` unnecessary.

Reply:
LinkedList is perhaps not as performant as you may think, but whatever -- I doubt it matter any much way.

I've been looking at your code a couple of times now and I finally get what you are doing. But the removeIf call has a very high "logic density". I think it would be better if you wrapped it in a method call, so you could do something like:
`checkAndRemovePending(entry -> entry.interaction.onSlotUpdate(containerId, slotNum, newStack));`

and have the expiry time check and the `!` inversion inside `checkAndRemovePending`.